lib: os: mpsc_pbuf: Test and Fix for concurrency issues

nordic-krch opened the pull request:
Added stress test for `mpsc_pbuf`.

Test is using 3 contexts to access single mpsc_pbuf instance. Those contexts
are on different priorities (2 threads and timer interrupt) and preempt
each other. One context is consuming and other two are producing. It
validates that each produced packet is consumed or dropped.

Test is randomized. Thread sleep time and timer timeout are random. Packet
size is also random. Dedicated work is used to fill a pool of random number
(generating random numbers is time consuming so it is decoupled from the main
test.
 
Test attempts to stress mpsc_pbuf but having as many preemptions as possible.
In order to achieve that CPU load is monitored periodically and if load is
to low then sleep/timeout time is reduced by reducing a factor that
is used to calculate sleep/timeout time (factor * random number). Test aims
to keep cpu load at ~80%. Some room is left for keeping random number pool
filled.

A report with test statistics is printed after the test. Statistics contain number of operations, preemptions and avarage CPU load during the test. Example output:
```
START - test_stress_preemptions_low_consumer
Produce preemption: 3253, consume preemptions: 4239
Claim total: 23633, missed: 3
Failed allocations: 0
Random numbers total:161223 missed:7
Avarage CPU load:67%, measurements:67
 PASS - test_stress_preemptions_low_consumer in 20.122 seconds
===================================================================
START - test_stress_preemptions_mid_consumer
Produce preemption: 79617, consume preemptions: 2640
Claim total: 28059, missed: 5
Failed allocations: 1
Random numbers total:160919 missed:1
Avarage CPU load:67%, measurements:67
 PASS - test_stress_preemptions_mid_consumer in 20.123 seconds
===================================================================
START - test_stress_preemptions_high_consumer
Produce preemption: 0, consume preemptions: 2048
Claim total: 31955, missed: 146
Failed allocations: 0
Random numbers total:108988 missed:0
Avarage CPU load:66%, measurements:67
 PASS - test_stress_preemptions_high_consumer in 20.123 seconds
```

Test revealed issues in the algorithm. Issues has been fixed and test now passes.

Additionally. converted buffer to use flag for indicating that buffer is full. This way whole buffer is dedicated to user data (previously one word was used to distinguish between empty and full state). Issue was reported here #38268

Fixes #38268.

API const correctness fixed in #38779

nordic-krch committed:
lib: os: mpsc_pbuf: Improve debugging

Improved debugging messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
lib: os: mpsc_pbuf: Use flag for buffer full indication

Use flag instead of word in the buffer. Using this method allows
to dedicate full buffer capacity for data.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: mpsc_pbuf: Adapt test after changes in the module

Adapt test to the new method of buffer full detection. New
method increased buffer capacity.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
lib: os: mpsc_pbuf: Fix concurrency issues

Fixed issues which were leading to failures when producing
and consuming is preempted at various stages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: mpsc_pbuf: Add new test case

Added test case for putting words when packet is claimed.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: mpsc_pbuf: Add stress test

Added stress test which validates proper behavior of mpsc_pbuf
when there are 2 producing contexts and dedicated consumer
context. Various configuration are tested with consumer having
the lowest, medium and the higher priority.

Test produces random size packets at random intervals. It validates
that each produced packet is consumed or dropped.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

carlescufi commented:
@dcpleung and @andyross any chance to get a review from you here?

dcpleung commented:
I don't know enough of this to do a proper review... :(

nashif commented:
@palchak-google  please review

cfriedt commented:
Great fix - thank you! I think I we were running into this periodically as well.  @nordic-krch - are you aware of whether this issue has always existed? E.g. do we need to backport to v2.7 and other releases?

nordic-krch commented:
@cfriedt that might be a good candidate for 2.7. However, first we need to get it to master :). I will resolve conflicts soon.
I'm aware that it is difficult to review. Hopefully someone will have mercy.

The main thing that is fixed is that content of the packet that is being dropped and passed to the notification function is valid. Without this fix, it could have been overwritten so it could not be trusted. Logging does not use this content in the notification callback (just bumps the counter) so logging stress test did show any issues there.

cfriedt commented on subsys/testsuite/ztest/Kconfig:
```diff
@@ -90,6 +90,7 @@ config ZTEST_PARAMETER_COUNT
 
 config ZTRESS
 	bool "Stress test framework"
+	depends on !SMP
```
`!SMP` seems somewhat counterintuitive - for stressing, it would make sense to run a number of threads competing for resources, wouldn't it?

nordic-krch replied:
need to remove it, when introduced ztress, initially it didn't run with SMP. It is fixed now. Will remove it.

cfriedt commented:
@nordic-krch - any chance this will be updated / rebased soon?

andyross commented on tests/lib/mpsc_pbuf/testcase.yaml:
```diff
@@ -5,3 +5,14 @@ tests:
+  lib.mpsc_pbuf_concurrent:
+    tags: mpsc_pbuf
+    platform_allow: >
+      qemu_cortex_m3 qemu_x86
```
Maybe add a SMP platform here, sub x86_64 for x86, for example?

nordic-krch replied:
add and luckily it revealed one bug.

andyross commented on tests/lib/mpsc_pbuf/testcase.yaml:
```diff
@@ -5,3 +5,14 @@ tests:
+    platform_allow: >
+      qemu_cortex_m3 qemu_x86
+    extra_configs:
+      - CONFIG_SYS_CLOCK_TICKS_PER_SEC=100000
```
What's this for?  A 1MHz tick rate isn't really achievable on any of our platforms, hardware or no.

andyross reviewed (commented):
Code looks fine.  A little concerned about the platform choices for coverage, and that tick rate has me very confused.

spiderkeys commented:
We've been testing these changes on our platform (Nucleo STM32L496ZG-P) for the last week or so and have not had any issues during runtime.

The test suite is also passing but there are some items that I wanted to double-check on. In the second run of each concurrency test, there are a large number of allocation failures (i.e. `Produced:18096, allocation failures:35103 Dropped: 0`, while the first test results look similar to those in the PR description. Is this normal/can you describe what these results mean?

```
Running TESTSUITE mpsc_pbuf_concurrent
===================================================================
START - test_stress_preemptions_high_consumer
11% remaining:0 ms ms
Ztress execution report:
         context 0:
                 - executed:18122, preempted:0
                 - ticks initial:20, optimized:4
         context 1:
                 - executed:17310, preempted:459
                 - ticks initial:20, optimized:4
         context 2:
                 - executed:16644, preempted:944
                 - ticks initial:20, optimized:4
         context 3:
                 - executed:15427, preempted:1790
                 - ticks initial:20, optimized:4
        Average CPU load:58%, measurements:100
Test report:
        Claims:18121, claim misses:1
        Produced:49381, allocation failures:0
        Dropped: 31251
0% remaining:0 ms ms
Ztress execution report:
         context 0:
                 - executed:18082, preempted:0
                 - ticks initial:20, optimized:4
         context 1:
                 - executed:18049, preempted:7
                 - ticks initial:20, optimized:4
         context 2:
                 - executed:17843, preempted:33
                 - ticks initial:20, optimized:4
         context 3:
                 - executed:17307, preempted:75
                 - ticks initial:20, optimized:4
        Average CPU load:55%, measurements:100
Test report:
        Claims:18081, claim misses:1
        Produced:18096, allocation failures:35103
        Dropped: 0
 PASS - test_stress_preemptions_high_consumer in 20.112 seconds
===================================================================
START - test_stress_preemptions_low_consumer
5% remaining:0 ms ms
Ztress execution report:
         context 0:
                 - executed:17661, preempted:0
                 - ticks initial:20, optimized:4
         context 1:
                 - executed:17290, preempted:227
                 - ticks initial:20, optimized:4
         context 2:
                 - executed:16501, preempted:988
                 - ticks initial:20, optimized:4
         context 3:
                 - executed:15280, preempted:1292
                 - ticks initial:20, optimized:4
        Average CPU load:56%, measurements:100
Test report:
        Claims:15280, claim misses:0
        Produced:51451, allocation failures:1
        Dropped: 36166
0% remaining:0 ms ms
Ztress execution report:
         context 0:
                 - executed:18745, preempted:0
                 - ticks initial:20, optimized:4
         context 1:
                 - executed:18473, preempted:19
                 - ticks initial:20, optimized:4
         context 2:
                 - executed:18170, preempted:45
                 - ticks initial:20, optimized:4
         context 3:
                 - executed:17234, preempted:170
                 - ticks initial:20, optimized:4
        Average CPU load:51%, measurements:100
Test report:
        Claims:17234, claim misses:0
        Produced:17250, allocation failures:38138
        Dropped: 0
 PASS - test_stress_preemptions_low_consumer in 20.112 seconds
===================================================================
START - test_stress_preemptions_mid_consumer
0% remaining:0 ms ms
Ztress execution report:
         context 0:
                 - executed:17362, preempted:0
                 - ticks initial:20, optimized:4
         context 1:
                 - executed:17322, preempted:7
                 - ticks initial:20, optimized:4
         context 2:
                 - executed:16992, preempted:313
                 - ticks initial:20, optimized:4
         context 3:
                 - executed:16034, preempted:1004
                 - ticks initial:20, optimized:4
        Average CPU load:58%, measurements:100
Test report:
        Claims:17320, claim misses:2
        Produced:50387, allocation failures:1
        Dropped: 33059
1% remaining:0 ms ms
Ztress execution report:
         context 0:
                 - executed:18889, preempted:0
                 - ticks initial:20, optimized:4
         context 1:
                 - executed:18149, preempted:47
                 - ticks initial:20, optimized:4
         context 2:
                 - executed:17960, preempted:71
                 - ticks initial:20, optimized:4
         context 3:
                 - executed:17433, preempted:77
                 - ticks initial:20, optimized:4
        Average CPU load:55%, measurements:100
Test report:
        Claims:18149, claim misses:0
        Produced:18157, allocation failures:36125
        Dropped: 0
 PASS - test_stress_preemptions_mid_consumer in 20.112 seconds
===================================================================
TESTSUITE mpsc_pbuf_concurrent succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

nordic-krch committed:
logging: Reset buffered message counter on init

Add reset of buffered messages counter to the initialization function.
It is for testing purposes sinces in the application logging is
initialized only once.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: logging: log_api: Align native posix case

Remove extra string size addition to the log message size when posix
is used. First string is considered as read-only also for posix so it
does not add to the message size. Test was passing before but change
in data handling in mpsc_pbuf revealed bug in the test.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: mpsc_pbuf: Add qemu_x86_64

Add target which supports SMP.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: mpsc_pbuf: Add native_posix to allowed list

Add plaftorm used for integration to platform_allowed list.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@peter-mitsis you previously approved. Since that i've added smp target to be used in the test (x86_64) and rebased.

nordic-krch commented:
@peter-mitsis ping

fabiobaltieri commented:
Hey, this (235ee632333164adf4d2342fe5258c9439d9dc75 specifically) is causing few failures in CI for `test_saturate`: https://github.com/zephyrproject-rtos/zephyr/runs/10348157924, seems mostly on qemu 64 architectures but there's qemu_xtensa there as well. @nordic-krch can you look into it?

nordic-krch commented:
@fabiobaltieri will be looking at it tomorrow.
